### ISSUE TITLE
`used-before-assignment` false positive for `type` alias

### ISSUE THREAD

**Issue by nickdrozd** [pylint-dev/pylint#9815]
### Bug description

```python
type IntOrX = int | X

class X:
    pass
```

### Command used

```shell
pylint --py-version=3.12 asdf.py
```

### Pylint output

```python
asdf.py:1:20: E0601: Using variable 'X' before assignment (used-before-assignment)
```

### Expected behavior

No warning

> The value of the type alias is evaluated in the annotation scope. It is not evaluated when the type alias is created, but only when the value is accessed through the type alias’s __value__ attribute (see [Lazy evaluation](https://docs.python.org/3/reference/executionmodel.html#lazy-evaluation)). This allows the type alias to refer to names that are not yet defined.

https://docs.python.org/3/reference/simple_stmts.html#the-type-statement

### Pylint version

```shell
pylint 3.3.0-dev0
astroid 3.3.0-dev0
Python 3.12.3
```

**Comment by nickdrozd** [comment]
No, it seems like it will be a lot of work. I am happy to help with PR review though.

Probably it is going to require a complete overhaul of the variables checker. I think it would be a good idea to fix this issue and https://github.com/pylint-dev/pylint/issues/9815 and other variable false positives (walrus, etc) all at once.
